lib/lf/uscalfe/fe_space_hp.h

```diff
@@ -0,0 +1,128 @@
+/**
+ * @file
+ * @brief Lagrangian finite elements of arbitrary polynomials degree
+ * @author Tobias Rohner
+ * @date   May 20202
+ * @copyright MIT License
+ */
+
+#ifndef LF_USCALFE_FE_SPACE_HP_H_
+#define LF_USCALFE_FE_SPACE_HP_H_
+
+#include <lf/mesh/utils/all_codim_mesh_data_set.h>
+#include "hp_fe.h"
+#include "scalar_fe_space.h"
+
+namespace lf::uscalfe {
```

Comment:
We have to discuss, what could a proper namespace for these hp-extensions.

Reply:
True. However, the way that @TobiasRohner has implemented it now is a natural generalization of uscalfe (i.e. UniformScalarFESpace inherits from ScalarFESpace) and it may be an option to rename `uscalfe -> fe`.

But that's only one of I think three options. 
1) If we are going to use ScalarFESpace quite often in the future (also useful e.g. for edge elements) I would keep ScalarFESpace in `lf::uscalfe`. The big advantage is that the ElementMatrix/VectorProviders can then be modified so that they work with ScalarFESpace and one can make the optimization for UniformScalarFESpace that we have right now. Also the MeshFunctionFE (and friends) do not need to be duplicated. This has the advantage, that the code must not be changed if one wants to switch from lagrange 2nd order to hp-spaces.
2) If we are going to rarely use ScalarFESpace we could put the classes in this PR into another namespace e.g. `lf::fe`. The advantage of this over 1) is that the `uscalfe` namespace is not polluted with classes that would almost never be used. The downside is however, that all the Matrix/Vector-Providers need to be reimplemented for ScalarFESpace as well as MeshFunctionFE (and related classes). Also it will not be so easy to switch from lagrangian spaces to hp-spaces.
3) If we never want to use the hp-spaces in another context, we could put the classes into the `examples` namespace.